openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,634 @@
+    fn from(diff: AbDiff<'a, T>) -> Self {
+        StagedAbDiff {
+            diff: diff.diff,
+            size: diff.size,
+        }
+    }
+}
+
+impl<T: Clone + Debug> StagedAbDiff<T> {
+    /// Return the actual diff inside of this StagedAbDiff
+    pub(super) fn diff(self) -> BTreeMap<NodeIndex, T> {
+        self.diff
+    }
+
+    /// Return the size of the diff.
+    pub(super) fn size(&self) -> TreeSize {
+        self.size
+    }
+}
+
+/// A [`NodeReference`] represents the position of a node in an [`AbDiff`]. It
+/// can be used to access the node at that position or to navigate to other,
+/// neighbouring nodes via the [`AbDiff::sibling()`], [`AbDiff::left_child()`]
+/// and [`AbDiff::right_child()`] functions of the [`AbDiff`].
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct NodeReference {
+    node_index: NodeIndex,
+}
+
+impl NodeReference {
+    /// Creates a new [`NodeReference`] to a node at the given index.
+    ///
+    /// Returns an error if the given index is outside the bounds of the diff.
+    pub(super) fn try_from_node_index<T: Clone + Debug>(
+        diff: &AbDiff<T>,
+        node_index: NodeIndex,
+    ) -> Result<Self, ABinaryTreeDiffError> {
+        diff.out_of_bounds(node_index)?;
+        Ok(NodeReference { node_index })
+    }
+}
+
+/// The [`AbDiff`] represents a set of differences (i.e. a "Diff") for an
+/// [`ABinaryTree`]. It can be created from an [`ABinaryTree`] instance and then
+/// accessed mutably or immutably. Any changes are saved by the [`AbDiff`] applied
+/// to the original [`ABinaryTree`] instance by converting it to a [`StagedAbDiff`]
+/// and subsequently merging it.
+pub(crate) struct AbDiff<'a, T: Clone + Debug> {
+    original_tree: &'a ABinaryTree<T>,
+    diff: BTreeMap<NodeIndex, T>,
+    size: TreeSize,
+}
+
+impl<'a, T: Clone + Debug> TryFrom<&'a ABinaryTree<T>> for AbDiff<'a, T> {
+    type Error = ABinaryTreeDiffError;
+
+    fn try_from(tree: &'a ABinaryTree<T>) -> Result<AbDiff<'a, T>, ABinaryTreeDiffError> {
+        Ok(AbDiff {
+            original_tree: tree,
+            diff: BTreeMap::new(),
+            size: tree.size()?,
+        })
+    }
+}
+
+impl<'a, T: Clone + Debug> AbDiff<'a, T> {
+    // Functions handling interactions with leaves.
+    ///////////////////////////////////////////////
+
+    /// Extend the diff by a leaf and its new parent node.
+    ///
+    /// Returns an error if adding either of the two nodes increases the size of
+    /// the diff beyond [`NodeIndex::MAX`].
+    pub(crate) fn add_leaf(
+        &mut self,
+        parent_node: T,
+        new_leaf: T,
+    ) -> Result<LeafIndex, ABinaryTreeDiffError> {
+        // Prevent the tree from becoming too large.
+        if self.size() >= NodeIndex::MAX - 1 {
+            return Err(ABinaryTreeDiffError::TreeTooLarge);
+        }
+        let original_size = self.size();
+        self.diff.insert(original_size, parent_node);
```

Comment:
`insert` returns a value if there has been one before. This shouldn't happen here but would be a good check for a `debug_assert`.
The same is true for the other operations on the diff.